File: minerva/qan.py
"""
QAN is a protocol that allows asking the peer a question and receiving an
answer or error-answer.  You can think of this as the "simplest wire-level
implementation of Deferreds", which is what AMP calls itself, though QAN takes
it even further: there is no command dispatching, and requests/responses
are completely unspecified.  (Though there are some helper functions here
that assume the requests/responses are bytestrings.)

QAN does not require Minerva; you can use it over any bidirectional framed
stream.
"""

import sys
import operator

from twisted.python import failure
from twisted.internet import defer
from twisted.python import log

from minerva.objcheck import strToNonNegLimit


_postImportVars = vars().keys()


# Not a meaningful superclass, just something to avoid copy/pasting
class _BodyAndId(tuple):
	__slots__ = ()

	body = property(operator.itemgetter(1))
	qid = property(operator.itemgetter(2))

	def __new__(cls, body, qid):
		return tuple.__new__(cls, (cls._MARKER, body, qid))


	def __repr__(self):
		return '%s(%r, %r)' % (self.__class__.__name__, self[1], self[2])



class _JustQid(tuple):
	__slots__ = ()

	qid = property(operator.itemgetter(1))

	def __new__(cls, body):
		return tuple.__new__(cls, (cls._MARKER, body))


	def __repr__(self):
		return '%s(%r)' % (self.__class__.__name__, self[1])



class _JustBody(tuple):
	__slots__ = ()

	body = property(operator.itemgetter(1))

	def __new__(cls, body):
		return tuple.__new__(cls, (cls._MARKER, body))


	def __repr__(self):
		return '%s(%r)' % (self.__class__.__name__, self[1])



class Question(_BodyAndId):
	__slots__ = ()
	_MARKER = object()



class OkayAnswer(_BodyAndId):
	__slots__ = ()
	_MARKER = object()



class KnownErrorAnswer(_BodyAndId):
	__slots__ = ()
	_MARKER = object()



class UnknownErrorAnswer(_BodyAndId):
	__slots__ = ()
	_MARKER = object()



class Cancellation(_JustQid):
	__slots__ = ()
	_MARKER = object()



class Notification(_JustBody):
	__slots__ = ()
	_MARKER = object()



qanTypeToCode = {
	 Question: "Q"
	,OkayAnswer: "K"
	,KnownErrorAnswer: "E"
	,UnknownErrorAnswer: "U"
	,Cancellation: "C"
	,Notification: "#"
}


def qanFrameToString(qanFrame):
	"""
	@param qanFrame: The QAN frame to encode
	@type qanFrame: a L{Question} or L{OkayAnswer} or L{KnownErrorAnswer} or
		L{UnknownErrorAnswer} or L{Cancellation} or L{Notification}, all of
		which must have a C{str} C{.body}.  (Except for L{Cancellation},
		which has no C{.body}.)

	@return: The encoded QAN frame
	@rtype: str
	"""
	qanFrameType = type(qanFrame)
	code = qanTypeToCode[qanFrameType]
	if qanFrameType == Cancellation:
		return str(qanFrame.qid) + code
	else:
		if not isinstance(qanFrame.body, str):
			raise TypeError("qanFrame.body must be a str, was %r" % (
				qanFrame.body,))

		if qanFrameType == Notification:
			return qanFrame.body + code
		else:
			return qanFrame.body + "|" + str(qanFrame.qid) + code


class InvalidQANFrame(Exception):
	pass



def _qidOrThrow(s):
	try:
		return strToNonNegLimit(s, 2**53)
	except ValueError:
		raise InvalidQANFrame("bad qid")


def stringToQANFrame(frameString):
	"""
	@param frameString: The QAN frame, encoded as a string
	@type frameString: C{str}

	@return: The QAN frame
	@rtype qf: a L{Question} or L{OkayAnswer} or L{ErrorAnswer} or
		L{Cancellation} or L{Notification}, with a C{str} C{.body}.
	"""
	try:
		lastByte = frameString[-1]
	except IndexError:
		raise InvalidQANFrame("0-length frame")

	if lastByte == "#":
		return Notification(frameString[:-1])
	elif lastByte == "C":
		qid = _qidOrThrow(frameString[:-1])
		return Cancellation(qid)
	else:
		try:
			body, rest = frameString.rsplit('|', 1)
		except ValueError:
			raise InvalidQANFrame("Expected pipe char in frame")
		qid = _qidOrThrow(rest[:-1])

		if lastByte == "Q":
			return Question(body, qid)
		elif lastByte == "K":
			return OkayAnswer(body, qid)
		elif lastByte == "E":
			return KnownErrorAnswer(body, qid)
		elif lastByte == "U":
			return UnknownErrorAnswer(body, qid)
		else:
			raise InvalidQANFrame("Invalid QAN frame type %r" % lastByte)


def isAnswerFrame(qanFrame):
	return isinstance(qanFrame, (OkayAnswer, KnownErrorAnswer, UnknownErrorAnswer))


class KnownError(Exception):
	pass



class UnknownError(Exception):
	pass



class QuestionFailed(Exception):
	pass



class QANHelper(object):
	def __init__(self, bodyReceived, logError, sendQANFrame, fatalError):
		"""
		@param bodyReceived: The 2-arg function to call when
			a Question or Notification is received (via a call to .handleString).
			Called with arguments: body, isQuestion.

		@param logError: A 2-arg function called when bodyReceived raises
			an exception.  Called with arguments: error message, L{Failure} object

		@param sendQANFrame: A 1-arg function that sends a QAN frame to the
			peer.  Called with argument: qanFrame.

		@param fatalError: A 1-arg function called when QANHelper can no longer
			handle QAN frames.  Called with argument: reason (a C{str} with bytes
			in inclusive range 0x20 (SPACE) to 0x7E (~)).
		"""
		self._bodyReceived = bodyReceived
		self._logError = logError
		self._sendQANFrame = sendQANFrame
		self._fatalError = fatalError

		self._qidCounter = 0
		self._ourQuestions = {}
		self._theirQuestions = {}


	def __repr__(self):
		return ('<%s asked %d questions, waiting for %d peer answers '
			'and %d local answers>') % (self.__class__.__name__,
				self._qidCounter, len(self._ourQuestions),
				len(self._theirQuestions))


	def _sendOkayAnswer(self, s, qid):
		del self._theirQuestions[qid]
		self._sendQANFrame(OkayAnswer(s, qid))


	def _sendErrorAnswer(self, failure, qid):
		del self._theirQuestions[qid]
		if failure.check(KnownError):
			self._sendQANFrame(KnownErrorAnswer(failure.value[0], qid))
		elif failure.check(defer.CancelledError):
			self._sendQANFrame(UnknownErrorAnswer("CancelledError", qid))
		else:
			self._logError("Peer's Question #%d caused uncaught "
				"exception" % (qid,), failure)
			# We intentionally do not reveal information about the
			# exception.
			self._sendQANFrame(UnknownErrorAnswer("Uncaught exception", qid))


	def handleQANFrame(self, qanFrame):
		if isAnswerFrame(qanFrame):
			qid = qanFrame.qid
			try:
				d = self._ourQuestions.pop(qid)
			except KeyError:
				self._fatalError("Received an answer with invalid qid: %d" % (qid,))
				return

			if d is None:
				# Ignore the answer to a question we cancelled or failAll'ed.
				pass
			elif isinstance(qanFrame, OkayAnswer):
				d.callback(qanFrame.body)
			elif isinstance(qanFrame, KnownErrorAnswer):
				d.errback(KnownError(qanFrame.body))
			elif isinstance(qanFrame, UnknownErrorAnswer):
				d.errback(UnknownError(qanFrame.body))
			else:
				raise RuntimeError("handleQANFrame bug")

		elif isinstance(qanFrame, Notification):
			try:
				self._bodyReceived(qanFrame.body, False)
			except Exception:
				self._logError("Peer's Notification caused uncaught "
					"exception", failure.Failure())

		elif isinstance(qanFrame, Question):
			qid = qanFrame.qid
			if qid in self._theirQuestions:
				self._fatalError("Received Question with duplicate qid: %d" % (qid,))
				return
			d = defer.maybeDeferred(
				self._bodyReceived, qanFrame.body, True)
			self._theirQuestions[qid] = d
			d.addCallbacks(
				self._sendOkayAnswer, self._sendErrorAnswer,
				callbackArgs=(qid,), errbackArgs=(qid,))
			d.addErrback(log.err)

		elif isinstance(qanFrame, Cancellation):
			qid = qanFrame.qid
			try:
				# We don't .pop() it here because a cancelled Deferred
				# still goes through the callback or errback chain, and
				# our _sendOkayAnswer and _sendErrorAnswer
				# deletes it from self._theirQuestions.
				d = self._theirQuestions[qid]
			except KeyError:
				# Cancellations for nonexistent questions are ignored.
				pass
			else:
				d.cancel()


	def _sendCancel(self, qid):
		self._ourQuestions[qid] = None

		# Note: when we cancel something, we still expect to get either
		# an OkayAnswer or ErrorAnswer from the peer, at least in the
		# typical case where the Stream does not reset.
		self._sendQANFrame(Cancellation(qid))

		# Because we don't call .callback or .errback in this canceller,
		# Deferred calls .errback(CancelledError()) for us.


	def ask(self, body):
		"""
		Send a Question to the peer.

		@param body: The question body
		@type body: *

		@return: a Deferred that will callback the response object, or errback
			with L{KnownError} or L{UnknownError}.
		@rtype: L{defer.Deferred}
		"""
		self._qidCounter += 1
		qid = self._qidCounter
		self._sendQANFrame(Question(body, qid))

		assert qid not in self._ourQuestions
		d = defer.Deferred(lambda _: self._sendCancel(qid))
		self._ourQuestions[qid] = d
		return d


	def notify(self, body):
		"""
		Send a Notification to the peer.

		@param body: The notification body
		@type body: *

		@return: None
		@rtype: C{None}
		"""
		self._sendQANFrame(Notification(body))


	def failAll(self, reason):
		"""
		Errback all of our questions with L{QuestionFailed}.

		@param reason: Reason for failing; used as the QuestionFailed
			exception message.
		@type reason: C{str}
		"""
		# .copy() because some buggy errback might .ask() a question
		for qid, d in self._ourQuestions.copy().iteritems():
			self._ourQuestions[qid] = None
			d.errback(QuestionFailed(reason))



try: from refbinder.api import bindRecursive
except ImportError: pass
else: bindRecursive(sys.modules[__name__], _postImportVars)

File: js_minerva/cw/net/qan.js
/**
 * @fileoverview QAN is a protocol that allows asking the peer a question and
 * receiving an answer or error-answer.  You can think of this as the "simplest
 * wire-level implementation of Deferreds", which is what AMP calls itself,
 * though QAN takes it even further: there is no command dispatching,
 * and requests/responses are completely unspecified.  (Though there are
 * some helper functions here that assume the requests/responses are
 * strings.)
 *
 * QAN does not require Minerva; you can use it over any bidirectional framed
 * stream.
 */

goog.provide('cw.net.stringToQANFrame');
goog.provide('cw.net.qanFrameToString');
goog.provide('cw.net.InvalidQANFrame');
goog.provide('cw.net.QANHelper');
goog.provide('cw.net.KnownError');
goog.provide('cw.net.UnknownError');
goog.provide('cw.net.QuestionFailed');

goog.require('goog.asserts');
goog.require('goog.debug.Error');
goog.require('goog.structs.Map');
goog.require('cw.deferred');
goog.require('cw.math');
goog.require('cw.record');
goog.require('cw.repr');
goog.require('cw.string');



/**
 * @param {*} body
 * @param {number} qid
 * @extends {cw.record.Record}
 * @constructor
 */
cw.net.Question = function(body, qid) {
	cw.record.Record.call(this, 'Question', [body, qid]);
	/** @type {*} */
	this.body = body;
	/** @type {number} */
	this.qid = qid;
};
goog.inherits(cw.net.Question, cw.record.Record);



/**
 * @param {*} body
 * @param {number} qid
 * @extends {cw.record.Record}
 * @constructor
 */
cw.net.OkayAnswer = function(body, qid) {
	cw.record.Record.call(this, 'OkayAnswer', [body, qid]);
	/** @type {*} */
	this.body = body;
	/** @type {number} */
	this.qid = qid;
};
goog.inherits(cw.net.OkayAnswer, cw.record.Record);



/**
 * @param {*} body
 * @param {number} qid
 * @extends {cw.record.Record}
 * @constructor
 */
cw.net.KnownErrorAnswer = function(body, qid) {
	cw.record.Record.call(this, 'KnownErrorAnswer', [body, qid]);
	/** @type {*} */
	this.body = body;
	/** @type {number} */
	this.qid = qid;
};
goog.inherits(cw.net.KnownErrorAnswer, cw.record.Record);



/**
 * @param {*} body
 * @param {number} qid
 * @extends {cw.record.Record}
 * @constructor
 */
cw.net.UnknownErrorAnswer = function(body, qid) {
	cw.record.Record.call(this, 'UnknownErrorAnswer', [body, qid]);
	/** @type {*} */
	this.body = body;
	/** @type {number} */
	this.qid = qid;
};
goog.inherits(cw.net.UnknownErrorAnswer, cw.record.Record);



/**
 * @param {number} qid
 * @extends {cw.record.Record}
 * @constructor
 */
cw.net.Cancellation = function(qid) {
	cw.record.Record.call(this, 'Cancellation', [qid]);
	/** @type {number} */
	this.qid = qid;
};
goog.inherits(cw.net.Cancellation, cw.record.Record);



/**
 * @param {*} body
 * @extends {cw.record.Record}
 * @constructor
 */
cw.net.Notification = function(body) {
	cw.record.Record.call(this, 'Notification', [body]);
	/** @type {*} */
	this.body = body;
};
goog.inherits(cw.net.Notification, cw.record.Record);



/**
 * @typedef {(
 * 	cw.net.Question|
 * 	cw.net.OkayAnswer|
 * 	cw.net.KnownErrorAnswer|
 * 	cw.net.UnknownErrorAnswer|
 * 	cw.net.Cancellation|
 * 	cw.net.Notification
 * 	)}
 */
cw.net.QANFrame;



/**
 * @param {!cw.net.QANFrame} frame
 * @return {string}
 * @private
 */
cw.net.qanTypeToCode_ = function(frame) {
	if(frame instanceof cw.net.Question) {
		return "Q";
	} else if(frame instanceof cw.net.OkayAnswer) {
		return "K";
	} else if(frame instanceof cw.net.KnownErrorAnswer) {
		return "E";
	} else if(frame instanceof cw.net.UnknownErrorAnswer) {
		return "U";
	} else if(frame instanceof cw.net.Cancellation) {
		return "C";
	} else if(frame instanceof cw.net.Notification) {
		return "#";
	} else {
		throw Error("qanTypeToCode bug");
	}
};


/**
 * @param {!cw.net.QANFrame} qanFrame The QAN frame to encode.
 * 	Any QANFrame with a {@code .body} must have a {@code string}
 * 	{@code .body}.
 * @return {string} Encoded QAN frame
 */
cw.net.qanFrameToString = function(qanFrame) {
	var code = cw.net.qanTypeToCode_(qanFrame);
	if(qanFrame instanceof cw.net.Cancellation) {
		return String(qanFrame.qid) + code;
	} else {
		if(!goog.isString(qanFrame.body)) {
			throw Error("qanFrame.body must be a string, was " +
				cw.repr.repr(qanFrame.body));
		}

		if(qanFrame instanceof cw.net.Notification) {
			return qanFrame.body + code;
		} else {
			return qanFrame.body + "|" + String(qanFrame.qid) + code;
		}
	}
};


/**
 * @param {string} message
 * @constructor
 * @extends {goog.debug.Error}
 */
cw.net.InvalidQANFrame = function(message) {
	goog.debug.Error.call(this);

	/**
	 * @type {string}
	 */
	this.message = message;
};
goog.inherits(cw.net.InvalidQANFrame, goog.debug.Error);



/**
 * @param {string} s A qid encoded as a string
 * @return {number} qid
 * @throws {cw.net.InvalidQANFrame} If {@code s} is not a valid qid
 * @private
 */
cw.net.qidOrThrow_ = function(s) {
	var n = cw.string.strToNonNegLimit(s, cw.math.LARGEST_INTEGER);
	if(goog.isNull(n)) {
		throw new cw.net.InvalidQANFrame("bad qid");
	}
	return n;
};


/**
 * @param {string} frameString The QAN frame, encoded as a string
 * @return {!cw.net.QANFrame}
 * @throws {cw.net.InvalidQANFrame} If frameString is not a valid QAN frame.
 */
cw.net.stringToQANFrame = function(frameString) {
	if(!frameString) {
		throw new cw.net.InvalidQANFrame("0-length frame");
	}

	var lastByte = frameString.substr(frameString.length - 1, 1);

	if(lastByte == "#") {
		return new cw.net.Notification(cw.string.withoutLast(frameString, 1));
	} else if(lastByte == "C") {
		var qid = cw.net.qidOrThrow_(cw.string.withoutLast(frameString, 1));
		return new cw.net.Cancellation(qid);
	} else {
		var _ = cw.string.rsplit(frameString, '|', 1);
		var body = _[0];
		var rest = _[1];
		if(!goog.isDef(rest)) {
			throw new cw.net.InvalidQANFrame("Expected pipe char in frame");
		}
		var qid = cw.net.qidOrThrow_(cw.string.withoutLast(rest, 1));

		if(lastByte == "Q") {
			return new cw.net.Question(body, qid);
		} else if(lastByte == "K") {
			return new cw.net.OkayAnswer(body, qid);
		} else if(lastByte == "E") {
			return new cw.net.KnownErrorAnswer(body, qid);
		} else if(lastByte == "U") {
			return new cw.net.UnknownErrorAnswer(body, qid);
		} else {
			throw new cw.net.InvalidQANFrame(
				"Invalid QAN frame type " + cw.repr.repr(lastByte));
		}
	}
};

/**
 * @param {!cw.net.QANFrame} qanFrame
 * @return {boolean} Is qanFrame an OkayAnswer, KnownErrorAnswer, or
 * 	UnknownErrorAnswer?
 */
cw.net.isAnswerFrame = function(qanFrame) {
	return (
		qanFrame instanceof cw.net.OkayAnswer ||
		qanFrame instanceof cw.net.KnownErrorAnswer ||
		qanFrame instanceof cw.net.UnknownErrorAnswer);
};


/**
 * @param {*} body
 * @constructor
 * @extends {goog.debug.Error}
 */
cw.net.KnownError = function(body) {
	goog.debug.Error.call(this);

	/**
	 * @type {*}
	 */
	this.body = body;
};
goog.inherits(cw.net.KnownError, goog.debug.Error);

/**
 * Message text.
 * @type {string}
 * @override
 */
cw.net.KnownError.prototype.message = 'KnownError with arbitrary body';

/**
 * @param {!Array.<string>} sb
 * @param {!Array.<*>} stack
 */
cw.net.KnownError.prototype.__reprPush__ = function(sb, stack) {
	sb.push("new KnownError(");
	cw.repr.reprPush(this.body, sb, stack);
	sb.push(")");
};



/**
 * @param {*} body
 * @constructor
 * @extends {goog.debug.Error}
 */
cw.net.UnknownError = function(body) {
	goog.debug.Error.call(this);

	/**
	 * @type {*}
	 */
	this.body = body;
};
goog.inherits(cw.net.UnknownError, goog.debug.Error);

/**
 * Message text.
 * @type {string}
 * @override
 */
cw.net.UnknownError.prototype.message = 'UnknownError with arbitrary body';

/**
 * @param {!Array.<string>} sb
 * @param {!Array.<*>} stack
 */
cw.net.UnknownError.prototype.__reprPush__ = function(sb, stack) {
	sb.push("new UnknownError(");
	cw.repr.reprPush(this.body, sb, stack);
	sb.push(")");
};



/**
 * @param {string} message
 * @constructor
 * @extends {goog.debug.Error}
 */
cw.net.QuestionFailed = function(message) {
	goog.debug.Error.call(this);

	/**
	 * @type {string}
	 */
	this.message = message;
};
goog.inherits(cw.net.QuestionFailed, goog.debug.Error);



/**
 * @param {function(*, boolean)} bodyReceived: The 2-arg function to call when
 * 	a Question or Notification is received (via a call to .handleString).
 * 	Called with arguments: body, isQuestion.
 *
 * @param {function(string, *)} logError: A 2-arg function called when
 * 	bodyReceived throws an error, or when QANHelper has a non-fatal
 * 	internal error.  Called with arguments: error message, error object.
 *
 * @param {function(!cw.net.QANFrame)} sendQANFrame: A 1-arg function that
 * 	sends a QAN frame to the peer.  Called with argument: qanFrame.
 *
 * @param {function(string)} fatalError: A 1-arg function called when QANHelper
 * 	can no longer handle QAN frames.  Called with argument: reason (a string
 * 	with chars in inclusive range 0x20 (SPACE) to 0x7E (~)).  After fatalError
 * 	is called, you must stop calling C{handleQANFrame} to prevent fatalError
 * 	from possibly being called again.
 *
 * @constructor
 */
cw.net.QANHelper = function(bodyReceived, logError, sendQANFrame, fatalError) {
	/**
	 * @type {function(*, boolean)}
	 * @private
	 */
	this.bodyReceived_ = bodyReceived;
	/**
	 * @type {function(string, *)}
	 * @private
	 */
	this.logError_ = logError;
	/**
	 * @type {function(!cw.net.QANFrame)}
	 * @private
	 */
	this.sendQANFrame_ = sendQANFrame;
	/**
	 * @type {function(string)}
	 * @private
	 */
	this.fatalError_ = fatalError;

	/**
	 * @type {number}
	 * @private
	 */
	this.qidCounter_ = 0;
	/**
	 * @type {!goog.structs.Map}
	 * @private
	 */
	this.ourQuestions_ = new goog.structs.Map();
	/**
	 * @type {!goog.structs.Map}
	 * @private
	 */
	this.theirQuestions_ = new goog.structs.Map();
};

/**
 * @param {!Array.<string>} sb
 * @param {!Array.<*>} stack
 */
cw.net.QANHelper.prototype.__reprPush__ = function(sb, stack) {
	sb.push("<QANHelper asked ", String(this.qidCounter_), " questions, " +
		"waiting for ", String(this.ourQuestions_.getCount()), " peer answers and ",
		String(this.theirQuestions_.getCount()), " local answers>");
};

/**
 * @param {*} body
 * @param {number} qid
 * @private
 */
cw.net.QANHelper.prototype.sendOkayAnswer_ = function(body, qid) {
	this.theirQuestions_.remove(qid);
	this.sendQANFrame_(new cw.net.OkayAnswer(body, qid));
};

/**
 * @param {*} error
 * @param {number} qid
 * @private
 */
cw.net.QANHelper.prototype.sendErrorAnswer_ = function(error, qid) {
	this.theirQuestions_.remove(qid);
	if(error instanceof cw.net.KnownError) {
		this.sendQANFrame_(new cw.net.KnownErrorAnswer(error.body, qid));
	} else if(error instanceof goog.async.Deferred.CancelledError) {
		this.sendQANFrame_(new cw.net.UnknownErrorAnswer("CancelledError", qid));
	} else {
		this.logError_("Peer's Question #" + qid +" caused uncaught exception", error);
		// We intentionally do not reveal information about the
		// exception.
		this.sendQANFrame_(new cw.net.UnknownErrorAnswer("Uncaught exception", qid));
	}
};

/**
 * @param {!cw.net.QANFrame} qanFrame
 */
cw.net.QANHelper.prototype.handleQANFrame = function(qanFrame) {
	if(cw.net.isAnswerFrame(qanFrame)) {
		var qid = qanFrame.qid;
		var d = this.ourQuestions_.get(qid);
		this.ourQuestions_.remove(qid);
		if(!goog.isDef(d)) {
			this.fatalError_("Received an answer with invalid qid: " + qid);
			return;
		}

		if(goog.isNull(d)) {
			// Ignore the answer to a question we cancelled or failAll'ed.
		} else if(qanFrame instanceof cw.net.OkayAnswer) {
			d.callback(qanFrame.body);
		} else if(qanFrame instanceof cw.net.KnownErrorAnswer) {
			d.errback(new cw.net.KnownError(qanFrame.body));
		} else if(qanFrame instanceof cw.net.UnknownErrorAnswer) {
			d.errback(new cw.net.UnknownError(qanFrame.body));
		} else {
			throw Error("handleQANFrame bug");
		}

	} else if(qanFrame instanceof cw.net.Notification) {
		try {
			this.bodyReceived_(qanFrame.body, false);
		} catch(e) {
			this.logError_("Peer's Notification caused uncaught " +
				"exception", e);
		}

	} else if(qanFrame instanceof cw.net.Question) {
		var qid = qanFrame.qid;
		if(this.theirQuestions_.containsKey(qid)) {
			this.fatalError_("Received Question with duplicate qid: " + qid);
			return;
		}
		var d = cw.deferred.maybeDeferred(
			this.bodyReceived_, [qanFrame.body, true]);
		this.theirQuestions_.set(qid, d);
		var that = this;
		d.addCallbacks(
			function(body) {
				that.sendOkayAnswer_(body, qid);
				return null;
			},
			function(error) {
				that.sendErrorAnswer_(error, qid);
				return null;
			}
		);
		d.addErrback(function(err) {
			this.logError_("Bug in QANHelper.sendOkayAnswer_ or sendErrorAnswer_", err);
			return null;
		});

	} else if(qanFrame instanceof cw.net.Cancellation) {
		var qid = qanFrame.qid;

		// We don't .remove(qid) here because a cancelled Deferred
		// still goes through the callback or errback chain, and
		// our sendOkayAnswer_ and sendErrorAnswer_
		// deletes it from this.theirQuestions_.
		var d = this.theirQuestions_.get(qid);
		if(goog.isDef(d)) {
			d.cancel();
		}
		// Cancellations for nonexistent questions are ignored.
	}
};

/**
 * @param {number} qid
 * @private
 */
cw.net.QANHelper.prototype.sendCancel_ = function(qid) {
	this.ourQuestions_.set(qid, null);

	// Note: when we cancel something, we still expect to get either
	// an OkayAnswer or *ErrorAnswer from the peer, at least in the
	// typical case where the Stream does not reset.
	this.sendQANFrame_(new cw.net.Cancellation(qid));

	// Because we don't call .callback or .errback in this canceller,
	// Deferred calls .errback(CancelledError()) for us.
};

/**
 * Send a Question to the peer.
 *
 * @param {*} body The question body
 * @return {!goog.async.Deferred} a Deferred that will callback the response
 * 	object, or errback with {@code KnownError} or {@code UnknownError} or
 * 	{@code QuestionFailed}.
 */
cw.net.QANHelper.prototype.ask = function(body) {
	this.qidCounter_ += 1;
	var qid = this.qidCounter_;
	this.sendQANFrame_(new cw.net.Question(body, qid));

	goog.asserts.assert(!this.ourQuestions_.containsKey(qid));
	var that = this;
	var d = new goog.async.Deferred(function() {
		that.sendCancel_(qid);
	});
	this.ourQuestions_.set(qid, d);
	return d;
};

/**
 * Send a Notification to the peer.
 *
 * @param {*} body The notification body
 */
cw.net.QANHelper.prototype.notify = function(body) {
	this.sendQANFrame_(new cw.net.Notification(body));
};

/**
 * Errback all of our questions with {@code QuestionFailed}.
 *
 * @param {string} reason Reason for failing; used as the
 * 	{@code QuestionFailed} error message.
 */
cw.net.QANHelper.prototype.failAll = function(reason) {
	var keys = this.ourQuestions_.getKeys();
	for(var i=0; i < keys.length; i++) {
		var d = this.ourQuestions_.get(keys[i]);
		// isDef just in case a very buggy errback mutated ourQuestions_
		if(goog.isDef(d)) {
			this.ourQuestions_.set(keys[i], null);
			d.errback(new cw.net.QuestionFailed(reason));
		}
	}
};
